Framework/Crystal/src/FindUBUsingIndexedPeaks.cpp

```diff
@@ -82,31 +118,151 @@ void FindUBUsingIndexedPeaks::exec() {
-            std::string(
-                "New UB will index %1d Peaks out of %1d with tolerance %5.3f")
-                .c_str(),
-            num_indexed, n_peaks, tolerance);
-    g_log.notice(std::string(logInfo));
+        for (Peak peak : peaks) {
+          if (peak.getRunNumber() == run) {
+            V3D hkl(peak.getIntHKL()); // ##### KEEP
+            V3D mnp(peak.getIntMNP());
+            if (IndexingUtils::ValidIndex(hkl, 1.0) ||
+                IndexingUtils::ValidIndex(mnp, 1.0)) {
+              run_q_vectors.push_back(peak.getQSampleFrame());
+              run_hkl_vectors.emplace_back(hkl[0], hkl[1], hkl[2]);
+              run_mnp_vectors.emplace_back(mnp[0], mnp[1], mnp[2]);
+            }
+          }
+        }
+
+        if (run_indexed < 3)
+          continue;
+
+        std::vector<double> rsigabc(7);
+        IndexingUtils::Optimize_6dUB(UB, modUB, run_hkl_vectors,
+                                     run_mnp_vectors, ModDim, run_q_vectors,
+                                     rsigabc, sigq);
+        OrientedLattice run_lattice;
+        run_lattice.setUB(UB);
+        run_lattice.setModUB(modUB);
+        run_lattice.setError(rsigabc[0], rsigabc[1], rsigabc[2], rsigabc[3],
+                             rsigabc[4], rsigabc[5]);
+        g_log.notice() << run_lattice << "\n";
+
+        if (!CrossTerm) {
+          double average_error = 0.;
+          IndexingUtils::CalculateMillerIndices(
+              UB, run_q_vectors, 1.0, run_fhkl_vectors, average_error);
+          for (size_t i = 0; i < run_indexed; i++) {
+            if (IndexingUtils::ValidIndex(run_fhkl_vectors[i], tolerance))
+              continue;
+
+            V3D fhkl(run_fhkl_vectors[i]);
+            if (run_mnp_vectors[i][0] != 0) {
+              fhkl -= run_lattice.getModVec(1) * run_mnp_vectors[i][0];
+              if (IndexingUtils::ValidIndex(fhkl, satetolerance)) {
+                sate_indexed++;
+                V3D errhkl = fhkl - run_hkl_vectors[i];
+                errhkl = V3D(fabs(errhkl[0]), fabs(errhkl[1]), fabs(errhkl[2]));
+                errorHKL1 += errhkl;
+              }
+            } else if (run_mnp_vectors[i][1] != 0) {
+              fhkl -= run_lattice.getModVec(2) * run_mnp_vectors[i][1];
+              if (IndexingUtils::ValidIndex(fhkl, satetolerance)) {
+                sate_indexed++;
+                V3D errhkl = fhkl - run_hkl_vectors[i];
+                errhkl = V3D(fabs(errhkl[0]), fabs(errhkl[1]), fabs(errhkl[2]));
+                errorHKL2 += errhkl;
+              }
+            } else if (run_mnp_vectors[i][2] != 0) {
+              fhkl -= run_lattice.getModVec(3) * run_mnp_vectors[i][2];
+              if (IndexingUtils::ValidIndex(fhkl, satetolerance)) {
+                sate_indexed++;
+                V3D errhkl = fhkl - run_hkl_vectors[i];
+                errhkl = V3D(fabs(errhkl[0]), fabs(errhkl[1]), fabs(errhkl[2]));
+                errorHKL3 += errhkl;
+              }
```

Comment:
No, I looked at this using PredictSatellitePeaks for data with 2 vectors.  Glad you asked because the CrossTerms were not done correctly so I fixed so it indexes predicted peaks correctly.  
````python
LoadIsawPeaks(Filename='/SNS/TOPAZ/IPTS-10003/shared/K2V3O8_satellite_5c_transformed/8071_combined.integrate', OutputWorkspace='peaks')
LoadIsawUB(InputWorkspace='peaks', Filename='/SNS/TOPAZ/IPTS-10003/shared/K2V3O8_satellite_5c_transformed/8071_combined.mat')
PredictSatellitePeaks(Peaks='peaks', SatellitePeaks='cross', MaxOrder=2, GetModVectorsFromUB=True, CrossTerms=True, IncludeAllPeaksInRange=True, WavelengthMin=1, WavelengthMax=2, MinDSpacing=0.5, MaxDSpacing=3)
FindUBUsingIndexedPeaks(PeaksWorkspace='cross', ToleranceForSatellite=0.050000000000000003)
````

Reply:
> Glad you asked because the satellites were not done for CrossTerms=True so I fixed so it indexes predicted peaks correctly.

So did you also add a unit test case that covers this?